### PR TITLE
HPCC-16110 Dedup with KEEP<>1 and RIGHT returns error message

### PR DESCRIPTION
**Pull request by shamser**
Signed-off-by: Shamser Ahmed <shamser.ahmed@lexisnexis.co.uk>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by shamser** [comment]
@ghalliday Please can you review

**Comment by AttilaVamos** [comment]
The errors are:
```
126. Test: dedup.ecl
    '<Exception><Code>3000</Code><Source>Roxie</Source><Message>assert(numToKeep == 1) failed - file: ccdserver.cpp, line 7241</Message></Exception>
    '
131. Test: dedupright.ecl
    '<Exception><Code>3000</Code><Source>Roxie</Source><Message>assert(numToKeep == 1) failed - file: ccdserver.cpp, line 7241</Message></Exception>
    '
132. Test: dedupwithbest.ecl
    '<Exception><Code>3000</Code><Source>Roxie</Source><Message>assert(numToKeep == 1) failed - file: ccdserver.cpp, line 7241</Message></Exception>
    '
599. Test: sqgroup.ecl ( version: multiPart=false )
    '<Exception><Code>3000</Code><Source>Roxie</Source><Message>assert(numToKeep == 1) failed - file: ccdserver.cpp, line 7241 (in Project 4)</Message></Exception>
    '
```

**Comment by ghalliday** [comment]
@shamser various regression tests are failing, and the code makes some attempt to implement keep(n).

The failures need addressing/commenting.

**Comment by shamser** [comment]
@ghalliday Please can you review again. Thanks.

**Comment by shamser** [comment]
@ghalliday Regression issue fixed.  Please can you  review

File: roxie/ccd/ccdserver.cpp [hunk 1]
```diff
@@ -7238,6 +7238,7 @@ class CRoxieServerDedupKeepRightActivity : public CRoxieServerDedupActivity
 
     virtual void start(unsigned parentExtractSize, const byte *parentExtract, bool paused)
     {
+        assertex(numToKeep <= 1);
```
Comment (ghalliday) [inline roxie/ccd/ccdserver.cpp]: If an error can be provoked by SENSIBLE user CODE then it should be caught and throw with a better error message.

Comment (ghalliday) [inline roxie/ccd/ccdserver.cpp]: This test is in the wrong place - the value will not be correctly initialised until *after* the base class start method is called.

**Review by ghalliday — CHANGES_REQUESTED** [review]
This should also have a test that triggers the problem (KEEP with a stored variable?)

**Comment by ghalliday** [comment]
@shamser back to you

**Pull request closed without merging**